From stable-bounces@linux.kernel.org  Tue Nov 15 04:32:56 2005
Date: Tue, 15 Nov 2005 13:32:36 +0100
From: Harald Welte <laforge@netfilter.org>
To: Stable Kernel <stable@kernel.org>
Cc: 
Subject: [PATCH] [NETFILTER] PPTP helper: Fix endianness bug in GRE key / CallID NAT
Return-Path: stable-bounces@linux.kernel.org
Received: from smtp.osdl.org (smtp.osdl.org [65.172.181.4])
	by shell0.pdx.osdl.net (8.13.1/8.11.6) with ESMTP id jAFCWur6030796;
	Tue, 15 Nov 2005 04:32:56 -0800
Received: from hera.kernel.org (hera.kernel.org [140.211.167.34])
	by smtp.osdl.org (8.12.8/8.12.8) with ESMTP id jAFCWtnO007694
	(version=TLSv1/SSLv3 cipher=EDH-RSA-DES-CBC3-SHA bits=168 verify=NO);
	Tue, 15 Nov 2005 04:32:55 -0800
Received: from hera.kernel.org (localhost [127.0.0.1])
	by hera.kernel.org (8.13.1/8.13.1) with ESMTP id jAFCWkDU006315;
	Tue, 15 Nov 2005 04:32:46 -0800
Received: from ganesha.gnumonks.org (ganesha.gnumonks.org [213.95.27.120])
	by hera.kernel.org (8.13.1/8.13.1) with ESMTP id jAFCWefT006262
	for <stable@kernel.org>; Tue, 15 Nov 2005 04:32:42 -0800
Received: from berligate.hmw-consulting.de ([83.236.178.202]
	helo=sunbeam.hmw-consulting.de)
	by ganesha.gnumonks.org with esmtpsa (TLS-1.0:RSA_AES_256_CBC_SHA:32)
	(Exim 4.50) id 1EbzzJ-0001uc-Og; Tue, 15 Nov 2005 13:32:39 +0100
Received: from hanuman.de.gnumonks.org ([192.168.100.10] ident=Debian-exim)
	by sunbeam.hmw-consulting.de with esmtp (Exim 4.54)
	id 1EbzzJ-0001T0-1y; Tue, 15 Nov 2005 13:32:37 +0100
Received: from laforge by hanuman.de.gnumonks.org with local (Exim 4.54)
	id 1EbzzI-0006B3-OL; Tue, 15 Nov 2005 13:32:36 +0100
Message-ID: <20051115133236.610397000@hanuman.de.gnumonks.org>
User-Agent: davem-sendpatch.sh v0.2
X-Spam-Score: 0.0 (/)
X-Virus-Scanned: ClamAV version 0.85, clamav-milter version 0.85 on localhost
X-Virus-Scanned: ClamAV version 0.85, clamav-milter version 0.85 on localhost
X-Virus-Status: Clean
X-BeenThere: stable@linux.kernel.org
X-Mailman-Version: 2.1.5
Precedence: list
List-Id: For maintainers of the stable Linux series <stable.linux.kernel.org>
List-Unsubscribe: <http://linux.kernel.org/mailman/listinfo/stable>,
	<mailto:stable-request@linux.kernel.org?subject=unsubscribe>
List-Archive: <http://linux.kernel.org/mailman/private/stable>
List-Post: <mailto:stable@linux.kernel.org>
List-Help: <mailto:stable-request@linux.kernel.org?subject=help>
List-Subscribe: <http://linux.kernel.org/mailman/listinfo/stable>,
	<mailto:stable-request@linux.kernel.org?subject=subscribe>
Sender: stable-bounces@linux.kernel.org
Errors-To: stable-bounces@linux.kernel.org
Received-SPF: none (domain of stable-bounces@linux.kernel.org does not designate permitted sender hosts)
X-Spam-Status: No, hits=0 required=5 tests=
X-Spam-Checker-Version: SpamAssassin 2.63-osdl_revision__1.55__
X-MIMEDefang-Filter: osdl$Revision: 1.127 $
X-Scanned-By: MIMEDefang 2.36
Status: RO
Content-Length: 1847
Lines: 46

This endianness bug slipped through while changing the 'gre.key' field in
the conntrack tuple from 32bit to 16bit.

None of my tests caught the problem, since the linux pptp client always has
'0' as call id / gre key.  Only windows clients actually trigger the bug.

Signed-off-by: Harald Welte <laforge@netfilter.org>
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 net/ipv4/netfilter/ip_nat_proto_gre.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- linux-2.6.14.2.orig/net/ipv4/netfilter/ip_nat_proto_gre.c
+++ linux-2.6.14.2/net/ipv4/netfilter/ip_nat_proto_gre.c
@@ -139,8 +139,8 @@ gre_manip_pkt(struct sk_buff **pskb,
 			break;
 		case GRE_VERSION_PPTP:
 			DEBUGP("call_id -> 0x%04x\n", 
-				ntohl(tuple->dst.u.gre.key));
-			pgreh->call_id = htons(ntohl(tuple->dst.u.gre.key));
+				ntohs(tuple->dst.u.gre.key));
+			pgreh->call_id = tuple->dst.u.gre.key;
 			break;
 		default:
 			DEBUGP("can't nat unknown GRE version\n");
